rpcserver.go

```diff
@@ -6745,6 +6751,222 @@ func (r *rpcServer) ListPermissions(_ context.Context,
 	}, nil
 }
 
+func (r *rpcServer) updateFeatures(updates []*lnrpc.UpdateFeatureAction) (*lnrpc.Op, error) {
+
+	ops := &lnrpc.Op{Entity: "features"}
+
+	fsets := make(map[feature.Set]*lnwire.RawFeatureVector)
+	sets := []feature.Set{feature.SetNodeAnn}
+
+	for _, set := range sets {
+		raw := r.server.featureMgr.GetRaw(set)
+
+		for _, update := range updates {
+			bit := lnwire.FeatureBit(update.FeatureBit)
+			switch update.Action {
+			case lnrpc.UpdateAction_ADD:
+				if !raw.IsSet(bit) && raw.SafeSet(bit) != nil {
+					raw.Set(bit)
+					ops.Actions = append(ops.Actions,
+						fmt.Sprintf("%s set in %s",
+							lnwire.Features[bit],
+							set.String()))
+				}
+			case lnrpc.UpdateAction_REMOVE:
+				if raw.IsSet(bit) {
+					raw.Unset(bit)
+					ops.Actions = append(ops.Actions,
+						fmt.Sprintf("%s unset in %s",
+							lnwire.Features[bit],
+							set.String()))
+				}
+			default:
+				return nil, fmt.Errorf("invalid feature update action: %v", update.Action)
+			}
+		}
+
+		// Ensure that all of our feature sets properly set any
+		// dependent features.
+		fv := lnwire.NewFeatureVector(raw, lnwire.Features)
+		err := feature.ValidateDeps(fv)
+		if err != nil {
+			return nil, fmt.Errorf("invalid feature set %s: %v",
+				set.String(), err)
+		}
+
+		fsets[set] = raw
+	}
+
+	r.server.featureMgr.SetFeatureVectors(fsets)
+
+	return ops, nil
+}
+
+func (r *rpcServer) updateAddresses(currentAddresses []net.Addr, updates []*lnrpc.UpdateAddressAction) ([]net.Addr, *lnrpc.Op, error) {
+	// net.Addr is not comparable so we cannot use the default map (map[net.Addr]struct{})
+	// so we have to use arrays and a helping function
+	findAddr := func(addr net.Addr, slice []net.Addr) bool {
+		for _, sAddr := range slice {
+			if sAddr.Network() == addr.Network() && sAddr.String() == addr.String() {
+				return true
+			}
+		}
+		return false
+	}
+
+	// Preallocate enough memory for both arrays
+	removeAddr := make([]net.Addr, 0, len(updates))
```

Comment:
in this case the slices in the function will be small enough to be able to simple pre allocate memory and save time